ICE - Segmentation fault after `unknown trait item`

## Segmentation fault after `unknown trait item`  

- You can view the same on [`godbolt`](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYgAnKUcAZPAZMADl3ACNMYgkAJlIAB1QFQjsGFzcPb3jE5IEAoNCWCKiuWKtMGxShAiZiAjT3Tx8yioEqmoI8kPDImMtq2vqMpv6OwK7CnpKASktUE2Jkdg4AUmiAZjAwZYBWACExWlQAdwgTBjNMdCmdgBFljQBBKgYAahYmQIgpl%2BWAdl37g9zKYbC99sRlmsAY9AQRiB8CC8AGKoVA/f6Al5Yl5oBjmF7BF4gF4mNbRSHQh6Y7HPF4AWS%2BFMBfzuMMeeBYcVoyNRL34xDBNXRlOxfNewS%2BwpZ1KxAHpZS9IsQSMTCJgWD8AGwaYLLbUvPAKF6GY0KBSoZB4JgES5vTAEBAYUgvI4IPDIBAvdCoTAKBhgDiI94ED0GggKGUveWi0VwhFajSQpzg0081CQ7B6xMw36sqm5jgzWicba8TwcLSkVCcABKZkR5vmix%2B6x4pAImkLMwA1iB1gA6SQabaa6IaLiatYlLySbb6TiSMudqucXgKEAadudmZwWAwRAoVCcuiRciUNDH%2BhRYDIZDEBQAWj9TDiCkdFiwADd3ZgAGp4TAjgAeTiRhODbGhaBte9KDCZcwkCGoAE9wN4BDmGIJCgLCbRyg7bheAvNhBCAhhaBQiteCwMITGAJwDnXAjSCwd4jHESjmLwYg8LwT9fWXTBVHKEwbVQ8hBEwYsONoPAwnhTCXCwZc4Q5VCZioAxgAUf9AJAsCmP4QQRDEdgpBkQRFBUdQON0Lh9EMYx630WT10gGZUDiWwBEY3hUD44hiDwLBXKgZg2BAS4cgYUhPzEEwljHaIeCmGZmi8zwIEcQZPDs/wxgKIo9ASJJ0uyorsnSzoCsmSxJJ41oRjKuy0sqEYqu6Yo%2BnaJqutqdqJmKVK5gWUyixLJcOOrDgXlUAAOTUH01SQXgAcScJwayEF4IDWjatvecxIm%2BCBcEIEgWynKZeHwrQUtIXttk3KTF1IFgQEkLx%2B1mkckonNZtW%2BrxNVIctKymtcNy3Sid33CAkDmAg4hEs8IAvOIT2IYJWCWUxzGQF4uH7TVZv7StIqIQL0D0QzhFEcQzJpyy1GXWzSCOeE4jU%2BcOFLEHlymoCRKRxFUCoGb5sW5bds27bpf2phDuIY6XEvSILq4K6odusaOBet7Hv7KQ1l%2BWbZv%2BtYvCkDRpFB3zV0sSGbq7e6%2B1%2BInfkt/7tkkX5fit4GpLWCawftp2YfgXcDzRjGUejq8QGIKQvE3L8fx04DQPLCC6Gg9cIDgjj0OQsSi8w7DcJsMSiMYAhSPI5dqNo%2BjaFoRi2xYhz2MrfBuIqPjGLJoTkBEpY20CG0pMrGS5OQxSlkrFS3oI9TNO0gCM/0iDZGM%2BnpEZpRmZsvt7KMBOnOnkL3M8lJGIfB872tTAHwIJDQMhG50N8/zKf4tzausdKDgGDOFcA0PQeV8gdXKiVFIPVipRX6oVZqdUWgMDaAMUBGRkEANau0RBNVzDdUwTlXqoxIEDQkENJso1ua81tiuaauMQwE0HJ9DQ21ToU3VprMOMwECYCYFgKIXwXZcA0E9BcvB9abnoeDB2m4w6kEjnDEACNhaxyPOjK8WNwocDmgtJaq11oyx2sY%2BWitNbkxIEFam286amT3rIJm1lKy6FiOzF8XMpJ0P5pwQWiMRIvFFi8Jh%2BNCYfVJttFWWi1arDJDw7cfCBFCMoD2CQ4juZ6xAI9Pmk17brgUYkl2aw1hE1HBoY2HtNS/EkJIWaPhA7Bzthwa624lGw1hkgOOp4KCo00RjBOXBfgZNTosdOeks68EgrnWC8FEKYRLvMrCOE8JVyPMRWuZEKLd0wDROiDExIdzYvPKiXEeL9wEkPEeYlx6SWXNPeSSE57KUCkvZKfBV7jMzmJGmO8HHmXkAfFxOgQBrBPo5cwzkwiXyrNfby9s/JKiCr/ERLV7CZWAT1CB4wkFZBgQIOBFUUj4M6mitBjViF6DJegshOKCEUvSCQwhfV8pQI1rMahlDaFNIYSE%2BsYT%2BzDMiSdfAXC4ka1adDJJgiegiN7NESQ/ZtjewtsOScI46leFmpkqR2SZG%2BJafIrWhZ2kR3hkLZGvTumY2xpwUJLCIlk1FdYqmdlfn2IkI4iyQKWagrZhzLx41ckhw4P44WQSxb2sJkK9hEBokYwumsBJUrSD8JlcInWWScmyPyY7IpvYuDbH7NEbYZJhwfQVVwCcXBtWNODc0yV2tubRB5XI3hMVIhJHsJIIAA%3D%3D)

---

### I tried this code from [`E0324`](https://doc.rust-lang.org/error_codes/E0324.html) 

```rust
#![allow(unused)]
fn main() {
struct Bar;

trait Foo {
    const N : u32;

    fn M();
}

impl Foo for Bar {
    fn N() {}
    // error: item `N` is an associated method, which doesn't match its
    //        trait `<Bar as Foo>`
}
}
```

### Meta

- What version of Rust GCC were you using, git sha 1bc8f08453e4db8a679dd95b9ee635aeb8eb1b07.
- gccrs (Compiler-Explorer-Build-gcc-ccac7ed46c2e81434714d74a86d58d2e5e1df7c8-binutils-2.40) 13.0.1 20230417 (experimental)



---


### Error output
- ICE - after `method is not a member of trait`


<details><summary>Backtrace</summary>
  <p>
  
  ```rust
  E0324.rs:12:5: error: method ‘N’ is not a member of trait ‘Foo’
   11 | impl Foo for Bar {
      |      ~~~
   12 |     fn N() {}
      |     ^~
crab1: internal compiler error: Segmentation fault
0x11e1ad3 crash_signal
	../../gccrs/gcc/toplev.cc:314
0x7f9bb844251f ???
	./signal/../sysdeps/unix/sysv/linux/x86_64/libc_sigaction.c:0
0xc0a644 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::length() const
	/usr/include/c++/11/bits/basic_string.h:927
0xc0a644 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
	/usr/include/c++/11/bits/basic_string.h:459
0xc0a644 Rust::Resolver::TraitItemReference::get_identifier[abi:cxx11]() const
	../../gccrs/gcc/rust/typecheck/rust-hir-trait-reference.cc:45
0xc12bd8 Rust::Resolver::TypeCheckItem::validate_trait_impl_block(Rust::HIR::ImplBlock&, Rust::TyTy::BaseType*, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:575
0xc13133 Rust::Resolver::TypeCheckItem::visit(Rust::HIR::ImplBlock&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:346
0xc0d500 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:49
0xc0d500 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:35
0xc2906f ???
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-stmt.h:36
0xc2914c Rust::Resolver::TypeCheckStmt::Resolve(Rust::HIR::Stmt*)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-stmt.cc:34
0xc21034 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:539
0xc1fce3 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr*)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-expr.cc:41
0xc1204d Rust::Resolver::TypeCheckItem::visit(Rust::HIR::Function&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:434
0xc0d500 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:49
0xc0d500 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check-item.cc:35
0xbd3b5b Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
	../../gccrs/gcc/rust/typecheck/rust-hir-type-check.cc:37
0xa712bb Rust::Session::compile_crate(char const*)
	../../gccrs/gcc/rust/rust-session-manager.cc:631
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

  ```
  
  </p>
</details>

- After fixing this issue, we need to add [`E0324`](https://doc.rust-lang.org/error_codes/E0324.html) error code in gccrs

--- a/gcc/rust/typecheck/rust-hir-type-check-item.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-item.cc
@@ -615,7 +615,8 @@ TypeCheckItem::validate_trait_impl_block (
 						   impl_item.get (), self,
 						   specified_bound,
 						   substitutions);
-	  trait_item_refs.push_back (trait_item_ref.get_raw_item ());
+	  if (!trait_item_ref.is_error ())
+	    trait_item_refs.push_back (trait_item_ref.get_raw_item ());
 	}
     }
 

--- a/gcc/testsuite/rust/compile/issue-2478.rs
+++ b/gcc/testsuite/rust/compile/issue-2478.rs
@@ -1,0 +1,16 @@
+#[lang = "sized"]
+pub trait Sized {}
+
+struct Bar;
+
+trait Foo {
+    const N: u32;
+
+    fn M();
+}
+
+impl Foo for Bar {
+    // { dg-error "missing N, M in implementation of trait .Foo." "" { target *-*-* } .-1 }
+    fn N() {}
+    // { dg-error "method .N. is not a member of trait .Foo." "" { target *-*-* } .-1 }
+}

--- a/gcc/testsuite/rust/compile/non_member_const.rs
+++ b/gcc/testsuite/rust/compile/non_member_const.rs
@@ -7,9 +7,7 @@ trait Foo {
 
 struct Bar;
 
-impl Foo for Bar {
+impl Foo for Bar {// { dg-error "missing N in implementation of trait .Foo." }
     const N : u32 = 0; // { dg-error "item .N. is an associated const, which does not match its trait .Foo." }
-    // error: item `N` is an associated const, which doesn't match its
-    //        trait `<Bar as Foo>`
 }
 }

